io: Refactoring with Classes (#329)

totlmstr opened the pull request:
This PR implements ~~mark two of~~ io refactoring. `io` now uses two classes (one for directories and one for files) as the baseline for interacting with files, combined with [using Better Enums](https://github.com/aantron/better-enums) for manipulation of paths. As a result, *any* Vita path can now be constructed and translated.

~~This is a Work-in-Progress because directories still need to be refactored (traversing a directory WILL cause errors), but testing is needed.~~

totlmstr commented:
Refactoring is now finished.

Needs testing!

pent0 commented:
Tell me when this got ready to review. If we let it here too long there might be more merge conflicts

totlmstr commented:
@pent0 PR is now ready for review. Testing gave the green light.

pent0 commented:
Nice document, I will give it more look later

pent0 commented on src/emulator/io/include/io/state.h:
```diff
@@ -41,27 +42,177 @@ typedef std::shared_ptr<_WDIR> DirPtr;
+        return dir_info.sys_loc.generic_path().wstring().c_str();
+    }
+#else
+    dirent *get_dir_ptr() const {
```
This broke the abstraction? Consider making interface and implementations model

pent0 replied:
point still stands

totlmstr commented on src/emulator/io/include/io/state.h:
```diff
@@ -17,13 +17,14 @@
+#include <dirent.h>
 #include <map>
-#include <memory>
+#include <miniz.h>
```
It turns out that `miniz.h` is used over in `host` for `load_self`, but the shared pointers for the zip are only used in `interface` at emulator. When looking over it, I feel as if the `load_self` function itself should be broken down into smaller functions and `miniz` removed from `host` as a dependency for cleanliness, but that is outside the scope of this PR.

pent0 replied:
Okay

totlmstr commented:
Please review again, @pent0.

Rinnegatamante commented on src/emulator/io/include/io/io.h:
```diff
@@ -17,27 +17,8 @@
+constexpr int SCE_ERROR_ERRNO_EEXIST = 0x80010011; // File exists
+constexpr int SCE_ERROR_ERRNO_EMFILE = 0x80010018; // Too many files are open
+constexpr int SCE_ERROR_ERRNO_EBADFD = 0x80010051; // File descriptor is invalid for this operation
+constexpr int SCE_ERROR_UNKNOWN = 0xFFFFFFFF; // Unknown Error
```
does this error really exists?

Rinnegatamante commented on src/emulator/modules/SceAppUtil/SceAppUtil.cpp:
```diff
@@ -149,40 +150,37 @@ EXPORT(int, sceAppUtilResetCookieWebBrowser) {
+        case SCE_APPUTIL_SAVEDATA_DATA_SAVE_MODE_FILE:
         default:
-            fd = open_file(host.io, file_path, SCE_O_WRONLY | SCE_O_CREAT, host.pref_path.c_str(), export_name);
+            fd = open_file(host.io, file_path, SCE_O_WRONLY, host.pref_path, export_name);
```
why SCE_O_CREAT disappeared?

Rinnegatamante commented on src/emulator/modules/SceAppUtil/SceAppUtil.cpp:
```diff
@@ -211,28 +206,19 @@ EXPORT(int, sceAppUtilSaveDataMount) {
-    slot_path += std::to_string(slotId);
-    slot_path += ".bin";
-    SceUID fd = open_file(host.io, slot_path, SCE_O_WRONLY | SCE_O_CREAT, host.pref_path.c_str(), export_name);
+    const auto fd = open_file(host.io, construct_slotparam_path(slotId), SCE_O_CREAT, host.pref_path, export_name);
```
ditto but for WRITE.

1whatleytay commented on src/emulator/io/src/state_functions.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <io/state.h>
+
+SceOff FileStats::read(void *input_data, const int element_size, const SceSize element_count) const {
+    if (!fs::exists(file_info.sys_loc))
```
I feel like it would be more appropriate to check the file pointer, since you created that right off the bat.

1whatleytay commented on src/emulator/io/src/state_functions.cpp:
```diff
@@ -0,0 +1,64 @@
+    if (!can_write_file())
+        return -1;
+
+    if (!fs::exists(file_info.sys_loc.parent_path()))
```
Why is this different from the checks in read and seek?

1whatleytay commented on src/emulator/modules/SceAppUtil/SceAppUtil.cpp:
```diff
@@ -149,40 +150,37 @@ EXPORT(int, sceAppUtilResetCookieWebBrowser) {
+}
+
+const char *construct_slotparam_path(const unsigned int data) {
+    return construct_savedata0_path(std::string("SlotParam_" + std::to_string(data)).c_str(), "bin");
```
Calling `c_str` on a string that's about to go out of scope is undefined behaviour, and I **think** that's what's happening here.

totlmstr replied:
That is actually resolved with [this](https://github.com/totlmstr/Vita3K/blob/io-paths/src/emulator/util/src/util.cpp#L161) function mentioned in `device::construct_normalized_path`.

1whatleytay replied:
Nice. Is the result being freed later though?

totlmstr replied:
I ended up remodifying the function because it was being freed. Now, it is a `vector<char>` (which is recommended).

1whatleytay commented on src/emulator/modules/SceLibKernel/SceLibKernel.cpp:
```diff
@@ -357,26 +351,26 @@ EXPORT(int, sceIoDevctlAsync) {
+    return open_dir(host.io, dir, host.pref_path, export_name);
 }
 
-EXPORT(int, sceIoDread, SceUID fd, emu::SceIoDirent *dir) {
```
I think `emu::` is correct here.

totlmstr replied:
I removed the `emu::` here in favor of `vita-headers`' definition of SceIoDirent.

1whatleytay replied:
I don't think you should. The only difference between the `emu::` struct and vita-header's struct is that the pointer `d_private` is of correct size (4 bytes instead of 8).

totlmstr replied:
Confirmed that up on Discord (wanted a second opinion), so I reverted the changes.

1whatleytay commented on src/emulator/modules/SceLibKernel/SceLibKernel.cpp:
```diff
@@ -1092,36 +1086,36 @@ EXPORT(int, sceKernelGetTimerTime) {
         }
 
-        char *data = new char[size];
+        auto *data = new char[static_cast<int>(size) + 1]; // null-terminated char array
```
Why cast here?

totlmstr replied:
Partially for logging consistency.

Other lines for size were already casted to int for RET_ERROR, so it would make sense to have this line with size as casted to int as well.

1whatleytay replied:
Okay :)

totlmstr commented on src/emulator/io/include/io/device.h:
```diff
@@ -0,0 +1,141 @@
+  * \param path The input path to be tested.
+  * \return The path's root as an enumeration. Otherwise, VitaIoDevice::_INVALID.
+  */
+inline VitaIoDevice get_device(const std::string &path) {
```
See above resolved with pent0.

Some functions were (later) moved for the compiler.

1whatleytay replied:
Ahh, are you talking about https://github.com/Vita3K/Vita3K/pull/505#discussion_r305556235? Why only some and not all?

totlmstr replied:
Part lessening clutter, part class definition.

I treated the `device` namespace as a pseudo-class for `VitaIoDevice` (like member functions). This pretty much meant having short, compiler-friendly, fast functions as `inline`, and having them as such will be beneficial in the long run since they are faster (e.g., having `get_device` as inline will be really helpful in the long run due to all of the device path logic it does).

1whatleytay replied:
I mean, the cost of calling the function is going to be just a few cycles, plus I don't know how often they are really going to be called. I don't think they really need to be inline imo. Up to you though. If you think it's good, it's good.

totlmstr commented on src/emulator/io/include/io/filesystem.h:
```diff
@@ -0,0 +1,73 @@
+
+// For opening Boost.Filesystem files, Boost returns wide strings for Windows, normal strings for other OS
+#ifdef WIN32
+const wchar_t *translate_open_mode(const int flags);
```
While I do think that this might be a better solution for the char situation we have here, this would be better suited for a different PR.

1whatleytay replied:
Do you want me to make a pr to your branch and change them? I don't think it's that much work.

1whatleytay commented:
Ahhh! I wish there was just a commit with all the fixes I could look at. I'll review tomorrow, but in the meantime can you fix conflicts so we can get appveyor running?

totlmstr commented:
[Despite being correct in the implementation](https://github.com/Vita3K/Vita3K/commit/7967ed20086f4568a9d706215ef5fcbdce24e703#diff-ce79b55e74ed0c96f7346b1396e79040R160) with `vector<char>`, I was getting unreliable results in further testing with the normalized path (garbage results, even though AppVeyor compiled it correctly and the path there was fine). As such, I removed the conversion in favor of just keeping the path as a string.

For a simple idea on how to create paths for `io`, [the now-rebased `context.cpp` provides one idea](https://github.com/Vita3K/Vita3K/pull/505/commits/660c43c75504e6648288eaa2f703d9e012d52829#diff-0d5d80e10f400f4abb7494fcb5c8fce5R291).

One more review from both @1whatleytay and @Rinnegatamante should be enough to merge and close this PR.

1whatleytay commented on src/emulator/io/src/io.cpp:
```diff
@@ -390,333 +288,236 @@ SceOff seek_file(SceUID fd, SceOff offset, int whence, IOState &io, const char *
-        }
-
-        statp->st_size = sb.st_size;
+    struct stat sb;
```
Why `struct` in C++? T_T

totlmstr replied:
This line is a carry-over from before my PR.

[IEEE specification for getting file status in POSIX](http://pubs.opengroup.org/onlinepubs/009695399/functions/stat.html).

1whatleytay replied:
That's fair but I'd rather you changed it now that you've found it. T_T

1whatleytay commented on src/emulator/io/src/io.cpp:
```diff
@@ -390,333 +288,236 @@ SceOff seek_file(SceUID fd, SceOff offset, int whence, IOState &io, const char *
-int stat_file_by_fd(IOState &io, const int fd, SceIoStat *statp, const char *pref_path, uint64_t base_tick, const char *export_name) {
-    assert(statp != NULL);
+int stat_file_by_fd(IOState &io, const SceUID fd, SceIoStat *statp, const std::string &pref_path, SceUInt64 base_tick, const char *export_name) {
+    assert(statp != nullptr);
```
I would probably write just `assert(!statp);` but if the original writer was doing this to be more clear I understand.

1whatleytay commented on src/emulator/io/src/state_functions.cpp:
```diff
@@ -0,0 +1,61 @@
+#include <io/state.h>
+
+SceOff FileStats::read(void *input_data, const int element_size, const SceSize element_count) const {
+    if (!get_file_pointer())
```
Why use the get method when you have access to `wrapped_file` directly?

1whatleytay commented on src/emulator/modules/SceAppUtil/SceAppUtil.cpp:
```diff
@@ -149,40 +150,37 @@ EXPORT(int, sceAppUtilResetCookieWebBrowser) {
+}
+
+static std::string construct_slotparam_path(const unsigned int data) {
+    return construct_savedata0_path(std::string("SlotParam_" + std::to_string(data)), "bin");
```
Is `std::string` here redundant?

1whatleytay commented on src/emulator/modules/SceAppUtil/SceAppUtil.cpp:
```diff
@@ -149,40 +150,37 @@ EXPORT(int, sceAppUtilResetCookieWebBrowser) {
+    return construct_savedata0_path(std::string("SlotParam_" + std::to_string(data)), "bin");
+}
+
+EXPORT(int, sceAppUtilSaveDataDataRemove, SceAppUtilSaveDataFileSlot *slot, emu::SceAppUtilSaveDataRemoveItem *files, unsigned int fileNum, SceAppUtilSaveDataMountPoint *mountPoint) {
```
Ahh! `emu::` missing here too! You can use psp2 header if the struct does not contain a pointer or a struct that contains a pointer (because the sizes will be wrong). SceAppUtilSaveDataFileSlot has a pointer :/

1whatleytay commented on src/emulator/modules/SceAppUtil/SceAppUtil.cpp:
```diff
@@ -149,40 +150,37 @@ EXPORT(int, sceAppUtilResetCookieWebBrowser) {
 }
 
-EXPORT(int, sceAppUtilSaveDataDataSave, emu::SceAppUtilSaveDataFileSlot *slot, emu::SceAppUtilSaveDataFile *files, unsigned int fileNum, SceAppUtilSaveDataMountPoint *mountPoint, SceSize *requiredSizeKB) {
+EXPORT(int, sceAppUtilSaveDataDataSave, SceAppUtilSaveDataFileSlot *slot, emu::SceAppUtilSaveDataFile *files, unsigned int fileNum, SceAppUtilSaveDataMountPoint *mountPoint, SceSize *requiredSizeKB) {
```
Here too.

1whatleytay commented on src/emulator/np/src/trophy/trp_parser.cpp:
```diff
@@ -19,6 +19,8 @@
 #include <np/trophy/trp_parser.h>
 #include <util/bytes.h>
 
+#include <cstring>
```
Why is this header included?

totlmstr replied:
`np` had a compile error (related to strnicmp, iirc) and wanted this include. It wasn't previously included before, so it was somewhat unexpected.

1whatleytay replied:
Weird. Okay.

pent0 commented:
Could you also take #548 . It's ok if you decline though, but this amy suit for this PR

1whatleytay commented on src/emulator/io/src/state_functions.cpp:
```diff
@@ -0,0 +1,61 @@
+#include <io/state.h>
+
+SceOff FileStats::read(void *input_data, const int element_size, const SceSize element_count) const {
+    if (!wrapped_file.get())
```
You don't even need to use `.get()` :)

Rinnegatamante commented on src/emulator/io/src/io.cpp:
```diff
@@ -357,30 +257,28 @@ int write_file(SceUID fd, const void *data, SceSize size, const IOState &io, con
 }
 
-SceOff seek_file(SceUID fd, SceOff offset, int whence, IOState &io, const char *export_name) {
-    assert((whence == SCE_SEEK_SET) || (whence == SCE_SEEK_CUR) || (whence == SCE_SEEK_END));
```
Quite sure when whence is incorrect, a different error than SCE_ERROR_ERRNO_EBADFD is returned.

pent0 commented:
Nice job.
